File: ios/MagicFormsMobile/Services/FormSubmitAPI.swift
import Foundation

enum FormSubmitAPI {
    private static let decoder = JSONDecoder()

    static func fetchSchema(token: String, formId: Int) async throws -> FormSchemaResponse {
        let url = APIConfig.url(path: "/api/v1/forms/\(formId)/schema/")
        let data = try await AuthAPI.getData(url: url, token: token)
        let response = try decoder.decode(FormSchemaResponse.self, from: data)
        if response.ok { return response }
        throw try AuthAPI.decodeError(from: data)
    }

    static func submit(
        token: String,
        formId: Int,
        textFields: [String: String],
        checkboxFields: [String: Bool],
        checklistFields: [String: [String]],
        files: [String: (data: Data, fileName: String, mimeType: String)]
    ) async throws -> FormSubmitResponse {
        let url = APIConfig.url(path: "/api/v1/forms/\(formId)/submit/")
        let data = try await AuthAPI.postMultipartForm(
            url: url,
            token: token,
            textFields: textFields,
            checkboxFields: checkboxFields,
            checklistFields: checklistFields,
            files: files
        )
        let response = try decoder.decode(FormSubmitResponse.self, from: data)
        if response.ok { return response }
        if let fieldErrors = response.fieldErrors, !fieldErrors.isEmpty {
            throw FormSubmitValidationError(
                message: response.message ?? "Please fix the highlighted fields.",
                fieldErrors: fieldErrors
            )
        }
        throw try AuthAPI.decodeError(from: data)
    }

    static func fetchRelatedSchema(token: String, accessToken: String) async throws -> FormSchemaResponse {
        let url = APIConfig.url(path: "/api/v1/related/\(accessToken)/schema/")
        let data = try await AuthAPI.getData(url: url, token: token)
        let response = try decoder.decode(FormSchemaResponse.self, from: data)
        if response.ok { return response }
        throw try AuthAPI.decodeError(from: data)
    }

    static func submitRelated(
        token: String,
        accessToken: String,
        textFields: [String: String],
        checkboxFields: [String: Bool],
        checklistFields: [String: [String]],
        files: [String: (data: Data, fileName: String, mimeType: String)]
    ) async throws -> FormSubmitResponse {
        let url = APIConfig.url(path: "/api/v1/related/\(accessToken)/submit/")
        let data = try await AuthAPI.postMultipartForm(
            url: url,
            token: token,
            textFields: textFields,
            checkboxFields: checkboxFields,
            checklistFields: checklistFields,
            files: files
        )
        let response = try decoder.decode(FormSubmitResponse.self, from: data)
        if response.ok { return response }
        if let fieldErrors = response.fieldErrors, !fieldErrors.isEmpty {
            throw FormSubmitValidationError(
                message: response.message ?? "Please fix the highlighted fields.",
                fieldErrors: fieldErrors
            )
        }
        throw try AuthAPI.decodeError(from: data)
    }
}

struct FormSubmitValidationError: LocalizedError {
    let message: String
    let fieldErrors: [String: [String]]
    var errorDescription: String? { message }
}

File: android/app/src/main/java/com/magicforms/mobile/data/api/FormSubmitApi.kt
package com.magicforms.mobile.data.api

import com.magicforms.mobile.data.models.ApiException
import com.magicforms.mobile.data.models.FormSchemaResponse
import com.magicforms.mobile.data.models.FormSubmitResponse
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.RequestBody.Companion.toRequestBody

class FormSubmitApi {
    private val json = Json { ignoreUnknownKeys = true; isLenient = true }
    private val authApi = AuthApi()

    fun fetchSchema(token: String, formId: Int): FormSchemaResponse {
        val data = authApi.get(ApiConfig.url("/api/v1/forms/$formId/schema/"), token)
        val response = json.decodeFromString<FormSchemaResponse>(data)
        if (response.ok) return response
        throw ApiException(message = "Could not load form.")
    }

    fun submit(
        token: String,
        formId: Int,
        textFields: Map<String, String>,
        checkboxFields: Map<String, Boolean>,
        checklistFields: Map<String, List<String>>,
        files: Map<String, Pair<ByteArray, String>>,
    ): FormSubmitResponse {
        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
        textFields.forEach { (key, value) ->
            builder.addFormDataPart(key, value)
        }
        checkboxFields.filter { it.value }.forEach { (key, _) ->
            builder.addFormDataPart(key, "on")
        }
        checklistFields.forEach { (key, values) ->
            values.forEach { value -> builder.addFormDataPart(key, value) }
        }
        val octet = "application/octet-stream".toMediaType()
        files.forEach { (key, pair) ->
            val (bytes, fileName) = pair
            builder.addFormDataPart(
                key,
                fileName,
                bytes.toRequestBody(octet),
            )
        }
        val data = authApi.postMultipart(
            ApiConfig.url("/api/v1/forms/$formId/submit/"),
            builder.build(),
            token,
        )
        val response = json.decodeFromString<FormSubmitResponse>(data)
        if (response.ok) return response
        if (!response.fieldErrors.isNullOrEmpty()) {
            throw FormSubmitValidationException(
                message = response.message ?: "Please fix the highlighted fields.",
                fieldErrors = response.fieldErrors,
            )
        }
        throw ApiException(message = response.message ?: "Submit failed.")
    }

    fun fetchRelatedSchema(token: String, accessToken: String): FormSchemaResponse {
        val data = authApi.get(ApiConfig.url("/api/v1/related/$accessToken/schema/"), token)
        val response = json.decodeFromString<FormSchemaResponse>(data)
        if (response.ok) return response
        throw ApiException(message = "Could not load form.")
    }

    fun submitRelated(
        token: String,
        accessToken: String,
        textFields: Map<String, String>,
        checkboxFields: Map<String, Boolean>,
        checklistFields: Map<String, List<String>>,
        files: Map<String, Pair<ByteArray, String>>,
    ): FormSubmitResponse {
        val builder = MultipartBody.Builder().setType(MultipartBody.FORM)
        textFields.forEach { (key, value) -> builder.addFormDataPart(key, value) }
        checkboxFields.filter { it.value }.forEach { (key, _) -> builder.addFormDataPart(key, "on") }
        checklistFields.forEach { (key, values) ->
            values.forEach { value -> builder.addFormDataPart(key, value) }
        }
        val octet = "application/octet-stream".toMediaType()
        files.forEach { (key, pair) ->
            val (bytes, fileName) = pair
            builder.addFormDataPart(key, fileName, bytes.toRequestBody(octet))
        }
        val data = authApi.postMultipart(
            ApiConfig.url("/api/v1/related/$accessToken/submit/"),
            builder.build(),
            token,
        )
        val response = json.decodeFromString<FormSubmitResponse>(data)
        if (response.ok) return response
        if (!response.fieldErrors.isNullOrEmpty()) {
            throw FormSubmitValidationException(
                message = response.message ?: "Please fix the highlighted fields.",
                fieldErrors = response.fieldErrors,
            )
        }
        throw ApiException(message = response.message ?: "Submit failed.")
    }
}

class FormSubmitValidationException(
    override val message: String,
    val fieldErrors: Map<String, List<String>>,
) : Exception(message)
